### DISCUSSION TITLE
Is there a language switcher example?

### DISCUSSION THREAD

**Discussion by peterkelm** [datenstrom/community#127] [Ask a question] [answered]
Referring to https://github.com/datenstrom/yellow/wiki/Language-configuration

At the bottom of the page the documentation refers to crafting a language switcher similar to using "custom navigation". However, the API documentation is really scarce how to accomplish that - e.g. "getLanguages()" is not documented. 

This is what I tried so far - php code only as the HTML is pretty obvious anyway

<pre><code>
    $languages = $yellow->pages->getLanguages()
    foreach($languages as $language):
        // check if this is the active language
        echo ($yellow->page->get("language") == $language) ? " class=\"active\"" : "" 

        // need the same page in another language --- HOW TO DO THIS?
        echo $yellow->page->getLocation() // THIS DOES NOT WORK

        // get language name from text string file
        echo $yellow->text->getTextHtml("languageDescription", $language)
    endforeach
</code></pre>

What is the intended way to implement a language switcher?

BTW: I wonder how Yellow determines which pages belong together as translations. (Other systems like Kirby keep track of this "relationship" by keeping all translations of a page in one folder.)

**Comment by markseuffert** [comment, marked as answer]
I added an example, you can now use the setting `Navigation: navigation-multilanguage`. 

To answer your questions: In the [Yellow API](https://datenstrom.se/yellow/help/api-for-developers) the function `$yellow->pages->multi()` returns a page collection with multiple languages. If you enable the multi language mode, then Yellow knows what pages belong together by looking at the directory structure. 

Here's a snippet for showing current page with translations:

```
<?php $pagesMultiLanguage = $yellow->pages->multi($yellow->page->location) ?>
<?php foreach(pagesMultiLanguage as $page): ?>
... // generate your HTML here
<?php endforeach ?>
```

Here's a snippet for home page in different languages:

```
<?php $pagesMultiLanguage = $yellow->pages->multi("/") ?>
<?php foreach(pagesMultiLanguage as $page): ?>
... // generate your HTML here
<?php endforeach ?>
```

Hope this helps. :smile:

**Comment by peterkelm** [comment]
Mark,

Yes, it does - helps a lot in fact.

I was tinkering with multi() already as that seemed to be the right direction anyway. However, in my "experiments" that call always returned zero results. Looking at the differences between my code and yours, the difference became evident:

You used <pre>$yellow->page->location</pre> as the argument to the multi() call, I used <pre>$yellow->page->getLocation()</pre>. Just by looking at the API function name I expected both (location vs. getLocation()) do do the same. However, after reading through core.php and the docs again it turns out that I tried to use an "absolute location" instead of a relative one. That subtle difference did not catch my attention before.

P.S.: For the benefit of others it might be good to reevaluate whether getLocation() is the right name for the implemented behavior.

**Comment by markseuffert** [comment]
Yes, that's right. Perhaps an argument like `getLocation($absoluteLocation = false)` would work too and remove the confusion between the two formats. Something to look at when doing an API redesign. Appreciate the feedback!

**Comment by peterkelm** [comment]
Mark, I'd prefer two separate API calls with clear names to one with an argument - even if there is a "sane" default. But that's just my 2ct.

**Comment by nogginfuel** [comment]
I haven't really played around with multiple languages yet, however, I am fine with either approach as long as it is documented and added to the wiki documentation in the appropriate place.

**Comment by wunderfeyd** [comment]
The argumented form would be perfect if a switch between absolute and relative location at one place in the code is needed. Is there any usecase for this? If not two separate function calls would be fine, with precise naming like `getAbsoluteLocation` and `getRelativeLocation`. When doing this one could consider to rename other functions begining with `getLocation` depending on what kind of location is processed there.

**Comment by markseuffert** [comment]
Yellow works with relative locations internally. There are some API functions that work with either relative or absolute location, like `$yellow->pages->find($location, $absoluteLocation = false)`. There are four API functions which stand out and return an absolute location, like `$yellow->page->getLocation()`.  The main reason to introduce absolute locations was to generate HTML links in snippets with less code. At the moments templates/snippets/plugins use `$yellow->page->location` to access the relative location and `$yellow->page->getLocation()` to access the absolute location. Not a good solution.

From my personal view, I would use relative locations everywhere and return absolute locations only when required. There would be no more confusion (or maybe less confusion), when locations are considered relative per default. As it's done inside the core. Here are some examples to illustrate what I mean:

Typical function call: `onRequest($serverScheme, $serverName, $base, $location, $fileName)`
Classes working together: `$yellow->pages->find($yellow->page->getLocation())`
Snippet for navigation: `<a href="<?php echo $yellow->page->getLocation(true) ?>">Text</a>`\

Here are the same examples, this time with different function names:

Typical function call: `onRequest($serverScheme, $serverName, $base, $location, $fileName)`
Classes working together: `$yellow->pages->find($yellow->page->getLocation())`
Snippet for navigation: `<a href="<?php echo $yellow->page->getAbsoluteLocation() ?>">Text</a>`\

Just my 2 cents. What do you guys prefer?

**Comment by wunderfeyd** [comment]
What about the `find(..., $absoluteLocation...)`, `path(..., $absoluteLocation...)` and `multi(..., $absoluteLocation...)`, it won't be changed I think. In this case I would tend to the `getLocation($absoluteLocation)` variant because it would fit better together and one can pass through this argument if someone wraps the function call.

**Comment by peterkelm** [comment]
It might be easiest for now to just document the difference between `$yellow->page->location` and `$yellow->page->getLocation()` properly and re-evaluate changes to the find/path/multi API calls there is continuous demand for further clarification. I'm a big fan of working with relative locations, so I can relate very well to your comments, Mark.

As an alternative I would propose using `absLocation()`instead of `getLocation()` to make it clear that this is an absolute location/path. It would make sense to me to introduce `location()` as a documented way to access the relative location of a page.

@All, your thoughts?

**Comment by markseuffert** [comment]
I updated the [Yellow API](https://datenstrom.se/yellow/help/api-for-developers) with notes about locations.
